include/swift/DWARFImporter/DWARFImporter.h

```diff
@@ -0,0 +1,175 @@
+class ASTContext;
+class CompilerInvocation;
+class ClangImporterOptions;
+class ClangModuleUnit;
+class ClangNode;
+class Decl;
+class DeclContext;
+class ImportDecl;
+class IRGenOptions;
+class LazyResolver;
+class ModuleDecl;
+class NominalTypeDecl;
+class TypeDecl;
+class VisibleDeclConsumer;
+enum class SelectorSplitKind;
+
+/// Class that imports Clang modules into Swift, mapping directly
+/// from Clang ASTs over to Swift ASTs.
+class DWARFImporter final : public ClangModuleLoader {
+  friend class ClangModuleUnit;
+
+public:
+  class Implementation;
+
+private:
+  Implementation &Impl;
+
+  DWARFImporter(ASTContext &ctx, const ClangImporterOptions &clangImporterOpts,
+                DependencyTracker *tracker);
+
+public:
+  /// \brief Create a new DWARF importer that can import a suitable DWARF
+  /// module into the given ASTContext.
+  ///
+  /// \param ctx The ASTContext into which the module will be imported.
+  /// The ASTContext's SearchPathOptions will be used for the DWARF importer.
+  ///
+  /// \param importerOpts The options to use for the DWARF importer.
+  ///
+  /// \param tracker The object tracking files this compilation depends on.
+  ///
+  /// \returns a new DWARF module importer, or null (with a diagnostic) if
+  /// an error occurred.
+  static std::unique_ptr<DWARFImporter>
+  create(ASTContext &ctx,
+         const ClangImporterOptions &importerOpts,
+         DependencyTracker *tracker = nullptr);
+
+  DWARFImporter(const DWARFImporter &) = delete;
+  DWARFImporter(DWARFImporter &&) = delete;
+  DWARFImporter &operator=(const DWARFImporter &) = delete;
+  DWARFImporter &operator=(DWARFImporter &&) = delete;
+
+  ~DWARFImporter();
+
+  /// \brief Check whether the module with a given name can be imported without
+  /// importing it.
+  ///
+  /// Note that even if this check succeeds, errors may still occur if the
+  /// module is loaded in full.
+  virtual bool canImportModule(std::pair<Identifier, SourceLoc> named) override;
+
+  /// Adds a new search path to the Clang CompilerInstance, as if specified with
+  /// -I or -F.
+  ///
+  /// \returns true if there was an error adding the search path.
+  bool addSearchPath(StringRef newSearchPath, bool isFramework,
+                     bool isSystem) override;
+  
+  /// \brief Import a module with the given module path.
+  ///
+  /// Clang modules will be imported using the Objective-C ARC dialect,
+  /// with all warnings disabled.
+  ///
+  /// \param importLoc The location of the 'import' keyword.
+  ///
+  /// \param path A sequence of (identifier, location) pairs that denote
+  /// the dotted module name to load, e.g., AppKit.NSWindow.
+  ///
+  /// \returns the module referenced, if it could be loaded. Otherwise,
+  /// emits a diagnostic and returns NULL.
+  virtual ModuleDecl *loadModule(
+                        SourceLoc importLoc,
+                        ArrayRef<std::pair<Identifier, SourceLoc>> path)
+                      override;
```

Comment:
Nitpick: LLVM indentation style prefers double-indenting here instead of right-aligning.